rsync-upgrade

```diff
@@ -0,0 +1,143 @@
+# the Free Software Foundation, either version 3 of the License,
+# or any later version.
+#
+# NethServer is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with NethServer.  If not, see COPYING.
+#
+#
+# This script connect to a remoe host and synchronize all data
+# to the local machine.
+#
+# Usage example:
+#     nethservice.nethesis.it 2222
+#
+
+function help
+{
+    echo "Usage: $0 [-u] <host> <ssh_port>"
+    echo
+    echo "       -h    Display this help message."
+    echo "       -u    Stop services on the source machine and perform the upgrade."
+    echo
+    exit 0
+}
+
+_term() {
+    rm -f $include_f
+    rm -f $exclude_f
+}
+
+sync_only=1
+
+while getopts ":h:u" opt; do
+  case ${opt} in
+    h )
+      help
+      ;;
+    u )
+      sync_only=0
+      ;;
+    \? )
+      echo "Invalid Option: -$OPTARG" 1>&2
+      exit 1
+      ;;
+  esac
+done
+shift $((OPTIND -1))
+
+target=$1
+port=${2:-22}
+
+if [ -z $target ]; then
+    help
+fi
+
+trap _term SIGTERM EXIT
+
+# Checking dependencies
+rpm -q nethserver-backup-data >/dev/null
+if [ $? -gt 0 ]; then
+    yum install nethserver-backup-data -y
+fi
+
+
+
+# Prepare temp files
+ssh_keys=/root/.ssh/id_rsa
+include_f=$(mktemp --suffix=.include)
+exclude_f=$(mktemp --suffix=.exclude)
+hostname=$(hostname -f)
+
+
+ssh_options=" -o UserKnownHostsFile=/dev/null -o ControlMaster=auto -o ControlPath=~/.ssh/master-%r@%h:%p -o ControlPersist=1h -o StrictHostKeyChecking=no -p $port -i $ssh_keys"
+ssh_cmd="ssh $ssh_options"
+
+if [ ! -f /root/.ssh/id_rsa ]; then
+    ssh-keygen -f $ssh_keys -t rsa -N ''
+    ssh-copy-id "root@$target -p $port"
+fi
+
+
+if [ $sync_only -eq 0 ]; then
+     # Prepare the remote machine
+     echo "Disabling access to services..."
+     $ssh_cmd root@$target 'shorewall stop'
+     echo "Creating configuration backup on $target..."
+     $ssh_cmd root@$target '/sbin/e-smith/backup-config'
+     echo "Executing pre-backup-data on $target..."
+     $ssh_cmd root@$target '/sbin/e-smith/signal-event pre-backup-data'
+else
+     echo
+     echo "Sync-only mode enabled: no action will be perfomed on the remote machine"
+     echo
+fi
+
+# Retrieve the list of files to be copied and excluded
+$ssh_cmd root@$target 'for f in /etc/backup-data.d/*.include; do cat $f; echo; done | sort | uniq | sed '/^$/d'' > $include_f
+$ssh_cmd root@$target 'for f in /etc/backup-data.d/*.exclude; do cat $f; echo; done | sort | uniq | sed '/^$/d'' > $exclude_f
+
+# Do not sync .ssh to not lose ssh config
+if [ $sync_only -eq 1 ]; then
+    echo "/root/.ssh" >> $exclude_f
+fi
+
+# Execute the sync
+rsync -azvr --delete --files-from=$include_f --exclude-from=$exclude_f -e "$ssh_cmd" root@$target:/ /
```

Comment:
trap SIGINT to exit cleanly

Ref https://github.com/NethServer/nethserver-mail-server/blob/master/migration/sync_maildirs.sh#L84